lib/processPbrMetallicRoughness.js

```diff
@@ -0,0 +1,774 @@
+                type: WebGLConstants.FLOAT_VEC3
+            };
+            vertexShader += 'attribute vec3 a_normal;\n';
+            vertexShader += 'varying vec3 v_normal;\n';
+            if (hasSkinning) {
+                vertexShaderMain += '  v_normal = u_normalMatrix * mat3(skinMat) * weightedNormal;\n';
+            } else {
+                vertexShaderMain += '  v_normal = u_normalMatrix * weightedNormal;\n';
+            }
+
+            fragmentShader += 'varying vec3 v_normal;\n';
+        }
+
+        // Read tangents if available
+        if (hasTangents) {
+            techniqueAttributes.a_tangent = 'tangent';
+            techniqueParameters.tangent = {
+                semantic: 'TANGENT',
+                type: WebGLConstants.FLOAT_VEC3
+            };
+            vertexShader += 'attribute vec3 a_tangent;\n';
+            vertexShader += 'varying vec3 v_tangent;\n';
+            vertexShaderMain += '  v_tangent = (u_modelViewMatrix * vec4(weightedTangent, 1.0)).xyz;\n';
+
+            fragmentShader += 'varying vec3 v_tangent;\n';
+        }
+
+        // Add texture coordinates if the material uses them
+        var v_texcoord;
+        if (hasTexCoords) {
+            techniqueAttributes.a_texcoord_0 = 'texcoord_0';
+            techniqueParameters.texcoord_0 = {
+                semantic: 'TEXCOORD_0',
+                type: WebGLConstants.FLOAT_VEC2
+            };
+
+            v_texcoord = 'v_texcoord_0';
+            vertexShader += 'attribute vec2 a_texcoord_0;\n';
+            vertexShader += 'varying vec2 ' + v_texcoord + ';\n';
+            vertexShaderMain += '  ' + v_texcoord + ' = a_texcoord_0;\n';
+
+            fragmentShader += 'varying vec2 ' + v_texcoord + ';\n';
+        }
+
+        // Add skinning information if available
+        if (hasSkinning) {
+            techniqueAttributes.a_joint = 'joint';
+            var attributeType = getShaderVariable(skinningInfo.type);
+            var webGLConstant = glslTypeToWebGLConstant(attributeType);
+
+            techniqueParameters.joint = {
+                semantic: 'JOINTS_0',
+                type: webGLConstant
+            };
+            techniqueAttributes.a_weight = 'weight';
+            techniqueParameters.weight = {
+                semantic: 'WEIGHTS_0',
+                type: webGLConstant
+            };
+
+            vertexShader += 'attribute ' + attributeType + ' a_joint;\n';
+            vertexShader += 'attribute ' + attributeType + ' a_weight;\n';
+        }
+
+        vertexShader += 'void main(void) {\n';
+        vertexShader += vertexShaderMain;
+        vertexShader += '}\n';
+
+        // Fragment shader lighting
+        fragmentShader += 'const float M_PI = 3.141592653589793;\n';
+
+        var lambertianDiffuse = '';
+        lambertianDiffuse += 'vec3 lambertianDiffuse(vec3 baseColor) {\n';
+        lambertianDiffuse += '  return baseColor / M_PI;\n';
+        lambertianDiffuse += '}\n\n';
```

Comment:
Doing a multiline string is usually cleaner. Also use 4 spaces of indentation throughout.

```
var lambertianDiffuse =
    'vec3 lambertianDiffuse(vec3 baseColor) {\n' +
    '    return baseColor / M_PI;\n' +
    '}\n\n';
```